doc/developer/design/20230225_compute_introspection_schema.md

```diff
@@ -0,0 +1,212 @@
+the case for each relation.
+
+Furthermore, sometimes the same logical value is reported by two
+relations, one reporting `clusterd`'s view of the world, and the other
+reporting `environmentd`'s. For example, `mz_compute_frontiers`
+vs. `mz_cluster_replica_frontiers`. However, the names do not make it
+clear what the distinction is. 
+
+#### Redundant `mz_` in names.
+
+Every relation's name begins with `mz_`. This is redundant, because
+anyone querying these relations probably already understand that they
+are using Materialize. Even if not, they are already in a namespace
+called `mz_internal`.
+
+### Suggested fixes for schema inconsistencies
+
+* All per-worker relations should have a corresponding "global" view
+  where the results are aggregated across workers in the way most
+  likely to be useful (for example, summing, taking the minimum, or
+  restricting to worker 0). Because these global views are more likely
+  to be useful than the underlying sources, they should have the basic
+  name (e.g., `mz_arrangement_sharing`). The underlying source should
+  have a name with `per_worker` inserted (e.g.,
+  `mz_per_worker_arrangement_sharing`). Furthermore, the per-worker
+  variant should be clearly less prominent in documentation (either at
+  the bottom of the page, or in a "see more" section that has to be
+  clicked to expand).
+* The expectations about per-worker relations should be clearly
+  documented (e.g., we should clearly say whether the values for each
+  worker are expected to be the same during normal operation).
+* All relations that report a simple count or sum should be views based on
+  an `_internal` variant that maintains the count or sum in the
+  Differential diff field. We may consider renaming this, since it
+  conflicts confusingly with the overall name `mz_internal`.
+* `raw` variants should be removed, and the underlying dataflows
+  should be in terms of intervals.
+* All histogram views or sources should have a per-bucket `sum`
+  column.
+* All per-replica sources and views should have their names prepended
+  with some string (perhaps `clusterd_`, though I'm open to
+  suggestions) that makes it obvious which ones they are. Furthermore,
+  the limitations of per-replica relations should be clearly
+  documented.
+* The `mz_` prefix should be removed from all names.
+
+### Possibly useful missing views
+
+#### Channel-to-operator mapping
+
+See [GitHub
+PR](https://github.com/MaterializeInc/materialize/pull/17825).
+
+#### Scope child to parent mapping
+
+The only way to understand the structure of scope nesting currently is
+to use the `mz_dataflow_addresses` relation and do some tedious list
+manipulation. A simple view of `(id, parent_id)` would make this much
+simpler.
+
+#### More views enriched with operator names, dataflow names, etc.
+
+(Not quite sure about this -- it's tricky to balance ease-of-use with
+not blowing up the set of documented relations)
+
```

Comment:
I believe @frankmcsherry has a laundry list of controller-maintained introspection views that would be nice to have. Is this design document the right place to begin gathering those, or are those best left to a future design document?

Reply:
One thing Frank mentioned on Slack:

> In particular, it is hard to get end-to-end lag (from sources all the way trhough to indexes) though "easy" to get lag for one dataflow (it's inputs to outputs).

We should definitely have this, if we think of a good way to actually compute it ... you need to know the transitive dependencies of each dataflow, which is fundamentally recursive, and we don't (yet) have recursive queries outside of unsafe mode. I'll check whether it's easy to get transitive dependencies in the code.